File: system-test/typescript/src/testkit/driver/port/shop/shop-driver.ts
import { Result } from '../../../common/result.js';
import {
  PlaceOrderRequest,
  PlaceOrderResponse,
  ViewOrderResponse,
  SystemError,
  PublishCouponRequest,
  BrowseCouponsResponse,
} from '../../../common/dtos.js';

export interface ShopDriver {
  goToShop(): Promise<Result<void, SystemError>>;
  placeOrder(request: PlaceOrderRequest): Promise<Result<PlaceOrderResponse, SystemError>>;
  cancelOrder(orderNumber: string): Promise<Result<void, SystemError>>;
  deliverOrder(orderNumber: string): Promise<Result<void, SystemError>>;
  viewOrder(orderNumber: string): Promise<Result<ViewOrderResponse, SystemError>>;
  publishCoupon(request: PublishCouponRequest): Promise<Result<void, SystemError>>;
  browseCoupons(): Promise<Result<BrowseCouponsResponse, SystemError>>;
  close(): Promise<void>;
}

File: system-test/java/src/main/java/com/optivem/shop/testkit/driver/port/shop/ShopDriver.java
package com.optivem.shop.testkit.driver.port.shop;

import com.optivem.shop.testkit.driver.port.shop.dtos.error.SystemError;
import com.optivem.shop.testkit.driver.port.shop.dtos.BrowseCouponsResponse;
import com.optivem.shop.testkit.driver.port.shop.dtos.PlaceOrderRequest;
import com.optivem.shop.testkit.driver.port.shop.dtos.PlaceOrderResponse;
import com.optivem.shop.testkit.driver.port.shop.dtos.PublishCouponRequest;
import com.optivem.shop.testkit.driver.port.shop.dtos.ViewOrderResponse;
import com.optivem.shop.testkit.common.Result;

public interface ShopDriver extends AutoCloseable {
    Result<Void, SystemError> goToShop();

    Result<PlaceOrderResponse, SystemError> placeOrder(PlaceOrderRequest request);

    Result<Void, SystemError> cancelOrder(String orderNumber);

    Result<Void, SystemError> deliverOrder(String orderNumber);

    Result<ViewOrderResponse, SystemError> viewOrder(String orderNumber);

    Result<Void, SystemError> publishCoupon(PublishCouponRequest request);

    Result<BrowseCouponsResponse, SystemError> browseCoupons();
}
